File: year2020/day3/solution.rb
tree = '#'

def positions_for_step(step, line_size) = 0.step(by: step).lazy.map { _1 % line_size }

# part one
input     = File.open('input.txt').readlines.map(&:chomp)
line_size = input.first.size
positions = positions_for_step(3, line_size)

input.zip(positions).select do |row, idx|
  row[idx] == tree
end.count

# part two
slopes = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]

result = slopes.map do |step, row|
  positions_for_step(step, line_size).then do |positions|
    input[(0..) % row].zip(positions).select do |r, i|
      r[i] == tree
    end.count
  end
end

p result.reduce(1, :*)

File: year2020/day3/solution.clj
(ns solution
  (:require [clojure.string :as s]))

(def tree \#)
(def input (-> (slurp "input.txt") (s/split #"\n")))
(def line-length (count (first input)))
(def start-pos 0)

;; part 1
(defn count-trees
  [by-step piste]
  (let [positions      (iterate #(mod (+ by-step %1) line-length) start-pos)
        entries-on-pos (map nth piste positions)
        trees          (filter #(= tree %1) entries-on-pos)]
    (count trees)))

(count-trees 3 input)


;; part 2
(def slopes [[1 1] [3 1] [5 1] [7 1] [1 2]])

(reduce *
        1
        (for [[right down] slopes
              :let         [piste (take-nth down input)]]
          (count-trees right piste)))
